connector/src/main/scala/com/datastax/spark/connector/datasource/CassandraCatalog.scala

```diff
@@ -0,0 +1,431 @@
+      .toArray
+  }
+
+  /**
+    * Since we only allow single depth keyspace identifiers in C*
+    * we always either return an empty list of namespaces or
+    * throw a NoSuchNamespaceException
+    */
+  override def listNamespaces(namespace: Array[String]): Array[Array[String]] = {
+    getKeyspaceMeta(connector, namespace) // Thorows no such namespace if namespace is not found
+    Array.empty[Array[String]]
+  }
+
+
+
+  override def createNamespace(namespace: Array[String], metadata: java.util.Map[String, String]): Unit = {
+    val ksMeta = metadata.asScala
+    checkNamespace(namespace)
+    if (getMetadata(connector).getKeyspace(fromInternal(namespace.head)).asScala.isDefined) throw new NamespaceAlreadyExistsException(s"${namespace.head} already exists")
+    connector.withSessionDo { session =>
+      val createStmt = SchemaBuilder.createKeyspace(namespace.head)
+      val replicationClass = ksMeta.getOrElse(ReplicationClass, throw new CassandraCatalogException(s"Creating a keyspace requires a $ReplicationClass DBOption for the replication strategy class"))
+      val createWithReplication = replicationClass.toLowerCase(Locale.ROOT) match {
+        case SimpleStrategy =>
+          val replicationFactor = ksMeta.getOrElse(ReplicationFactor,
+            throw new CassandraCatalogException(s"Need a $ReplicationFactor option with SimpleStrategy"))
+          createStmt.withSimpleStrategy(replicationFactor.toInt)
+        case NetworkTopologyStrategy =>
+          val datacenters = (ksMeta -- IgnoredReplicationOptions).map(pair => (pair._1, pair._2.toInt: java.lang.Integer))
+          createStmt.withNetworkTopologyStrategy(datacenters.asJava)
+        case other => throw new CassandraCatalogException(s"Unknown keyspace replication strategy $other")
+      }
+      val finalCreateStmt = createWithReplication.withDurableWrites(ksMeta.getOrElse(DurableWrites, "True").toBoolean)
+      session.execute(finalCreateStmt.asCql())
+    }
+  }
+
+  override def alterNamespace(namespace: Array[String], changes: NamespaceChange*): Unit = {
+    checkNamespace(namespace)
+
+    val ksMeta: mutable.Map[String, String] = changes.foldRight(loadNamespaceMetadata(namespace).asScala) {
+      case (setProperty: SetProperty, metadata: mutable.Map[String, String]) =>
+        metadata + (setProperty.property() -> setProperty.value)
+      case (removeProperty: RemoveProperty, metadata: mutable.Map[String, String]) =>
+        metadata - removeProperty.property()
+      case (other, _) => throw new CassandraCatalogException(s"Unable to handle alter namespace operation: ${other.getClass.getSimpleName}")
+    }
+
+    val alterStart = SchemaBuilder.alterKeyspace(namespace.head)
+    val alterWithDurable = alterStart.withDurableWrites(ksMeta.getOrElse(DurableWrites, "True").toBoolean)
+    val replicationClass = ksMeta
+      .getOrElse(ReplicationClass, throw new CassandraCatalogException(s"Creating a keyspace requires a $ReplicationClass option"))
```

Comment:
The exception thrown here should probably reference altering a keyspace rather than creating one... maybe something like "cannot alter keyspace because replication class couldn't be determined" or something along those lines.

Reply:
Yep good call